web/app/Common/PGError.php

```diff
@@ -1,8 +1,196 @@
+    public const NUMERIC_VALUE_OUT_OF_RANGE = '22003';
+    public const STRING_DATA_LENGTH_MISMATCH = '22026';
+    public const STRING_DATA_RIGHT_TRUNCATION_22001 = '22001';
+    public const SUBSTRING_ERROR = '22011';
+    public const TRIM_ERROR = '22027';
+    public const UNTERMINATED_C_STRING = '22024';
+    public const ZERO_LENGTH_CHARACTER_STRING = '2200F';
+    public const FLOATING_POINT_EXCEPTION = '22P01';
+    public const INVALID_TEXT_REPRESENTATION = '22P02';
+    public const INVALID_BINARY_REPRESENTATION = '22P03';
+    public const BAD_COPY_FILE_FORMAT = '22P04';
+    public const UNTRANSLATABLE_CHARACTER = '22P05';
+    public const INTEGRITY_CONSTRAINT_VIOLATION = '23000';
+    public const RESTRICT_VIOLATION = '23001';
+    public const NOT_NULL_VIOLATION = '23502';
+    public const FOREIGN_KEY_VIOLATION = '23503';
+    public const UNIQUE_VIOLATION = '23505';
+    public const CHECK_VIOLATION = '23514';
+    public const INVALID_CURSOR_STATE = '24000';
+    public const INVALID_TRANSACTION_STATE = '25000';
+    public const ACTIVE_SQL_TRANSACTION = '25001';
+    public const BRANCH_TRANSACTION_ALREADY_ACTIVE = '25002';
+    public const HELD_CURSOR_REQUIRES_SAME_ISOLATION_LEVEL = '25008';
+    public const INAPPROPRIATE_ACCESS_MODE_FOR_BRANCH_TRANSACTION = '25003';
+    public const INAPPROPRIATE_ISOLATION_LEVEL_FOR_BRANCH_TRANSACTION = '25004';
+    public const NO_ACTIVE_SQL_TRANSACTION_FOR_BRANCH_TRANSACTION = '25005';
+    public const READ_ONLY_SQL_TRANSACTION = '25006';
+    public const SCHEMA_AND_DATA_STATEMENT_MIXING_NOT_SUPPORTED = '25007';
+    public const NO_ACTIVE_SQL_TRANSACTION = '25P01';
+    public const IN_FAILED_SQL_TRANSACTION = '25P02';
+    public const INVALID_SQL_STATEMENT_NAME = '26000';
+    public const TRIGGERED_DATA_CHANGE_VIOLATION = '27000';
+    public const INVALID_AUTHORIZATION_SPECIFICATION = '28000';
+    public const DEPENDENT_PRIVILEGE_DESCRIPTORS_STILL_EXIST = '2B000';
+    public const DEPENDENT_OBJECTS_STILL_EXIST = '2BP01';
+    public const INVALID_TRANSACTION_TERMINATION = '2D000';
+    public const SQL_ROUTINE_EXCEPTION = '2F000';
+    public const FUNCTION_EXECUTED_NO_RETURN_STATEMENT = '2F005';
+    public const MODIFYING_SQL_DATA_NOT_PERMITTED = '2F002';
+    public const PROHIBITED_SQL_STATEMENT_ATTEMPTED = '2F003';
+    public const READING_SQL_DATA_NOT_PERMITTED = '2F004';
+    public const INVALID_CURSOR_NAME = '34000';
+    public const EXTERNAL_ROUTINE_EXCEPTION = '38000';
+    public const CONTAINING_SQL_NOT_PERMITTED = '38001';
+    public const MODIFYING_SQL_DATA_NOT_PERMITTED_38002 = '38002';
+    public const PROHIBITED_SQL_STATEMENT_ATTEMPTED_38003 = '38003';
+    public const READING_SQL_DATA_NOT_PERMITTED_38004 = '38004';
+    public const EXTERNAL_ROUTINE_INVOCATION_EXCEPTION = '39000';
+    public const INVALID_SQLSTATE_RETURNED = '39001';
+    public const NULL_VALUE_NOT_ALLOWED_39004 = '39004';
```

Comment:
Why do some of these end in numbers?

Reply:
They are duplicate keys otherwise, they have the same error code twice in here, and I'm not familiar enough with the particulars of these errors to come up with a better name. We'll have to see if they're ever relevant and then maybe we refactor around it.

Reply:
Oh that's strange. Well this workaround makes sense to me: you may want to consider adding a comment like above to explain it to future versions of ourselves.